src/gluonnlp/lr_scheduler/reduce_lr_on_plateau.py

```diff
@@ -0,0 +1,210 @@
+                 trainer,
+                 mode='min',
+                 factor=0.1,
+                 patience=10,
+                 verbose=False,
+                 threshold=1e-4,
+                 threshold_mode='rel',
+                 cooldown=0,
+                 min_lr=0,
+                 eps=1e-8):
+
+        if factor >= 1.0:
+            raise ValueError('Factor should be < 1.0.')
+        self.factor = factor
+
+        if not isinstance(trainer, gluon.Trainer):
+            raise TypeError('{} is not an mxnet.trainer.trainer'.format(
+                type(trainer).__name__))
+        self.trainer = trainer
+
+        self.min_lr = min_lr
+
+        self.patience = patience
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0
+        self.mode = mode
+        self.threshold = threshold
+        self.threshold_mode = threshold_mode
+        self.best = None
+        self.num_bad_epochs = None
+        self.mode_worse = None  # the worse value for the chosen mode
+        self.is_better = None
+        self.eps = eps
+        self.last_epoch = -1
+        self._init_is_better(mode=mode,
+                             threshold=threshold,
+                             threshold_mode=threshold_mode)
+        self._reset()
+
+    def _reset(self):
+        r"""Resets num_bad_epochs counter and cooldown counter."""
+        self.best = self.mode_worse
+        self.cooldown_counter = 0
+        self.num_bad_epochs = 0
+
+    def step(self, metric, epoch=None):
+        r"""Function to be executed after model evaluation
+
+        Parameters
+        ----------
+        metric : float
+            Current metric value to mesure model performance.
+        epoch : int, default None
+            Current epoch. If None, it is managed internally.
+
+        """
+        current = float(metric)
+        if epoch is None:
+            epoch = self.last_epoch = self.last_epoch + 1
+        self.last_epoch = epoch
+
+        if self.is_better(current, self.best):
+            self.best = current
+            self.num_bad_epochs = 0
+        else:
+            self.num_bad_epochs += 1
+
+        if self.in_cooldown:
+            self.cooldown_counter -= 1
+            self.num_bad_epochs = 0
+
+        if self.num_bad_epochs > self.patience:
+            self._reduce_lr(epoch)
+            self.cooldown_counter = self.cooldown
+            self.num_bad_epochs = 0
+
+    def _reduce_lr(self, epoch):
+        old_lr = float(self.trainer.learning_rate)
+        new_lr = max(old_lr * self.factor, self.min_lr)
+        if old_lr - new_lr > self.eps:
+            self.trainer.set_learning_rate(new_lr)
+            if self.verbose:
+                print('Epoch {:5d}: reducing learning rate'
+                      ' {} to {}.'.format(epoch, old_lr, new_lr))
+
+    @property
+    def in_cooldown(self):
+        return self.cooldown_counter > 0
+
+    def _cmp(self, mode, threshold_mode, threshold, a, best):
```

Comment:
The current design does not look very scalable. That means it requires hard coding if we would like add some changes/schedules.

Reply:
Thanks for the comments.
I can consider more flexible class desine.